### PR TITLE
Directory search join button

### PR DESCRIPTION
**Pull request by dbkr**
Implement 'join' button that appears in room directory search box when you type a valid(ish) alias.

Also factors out the directory search box to a different component now it's a lot more complex (and also adds the clear button that was in the designs).

Fixes https://github.com/vector-im/vector-web/issues/2217

### CODE REVIEW

File: src/components/views/elements/DirectorySearchBox.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        }
+    }
+
+    onJoinButtonClick() {
+        if (this.props.onJoinClick) {
+            this.props.onJoinClick(this.state.value);
+        }
+    }
+
+    _contentLooksLikeAlias() {
```
Comment (richvdh) [inline src/components/views/elements/DirectorySearchBox.js]: what's with some of the internal functions having `_`, and some not? I'd prefer it if they all had _, but at least be consistent.

File: src/components/views/elements/DirectorySearchBox.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if (!this.input) return false;
+
+        // liberal test for things that look like room aliases
+        return /#.+:.+/.test(this.state.value);
```
Comment (richvdh) [inline src/components/views/elements/DirectorySearchBox.js]: this matches `abc#def:abc`. You probably want `/^#.+:/`

File: src/components/views/elements/DirectorySearchBox.js (hunk 1 above)
Comment (dbkr) [inline src/components/views/elements/DirectorySearchBox.js]: The logic, ftr, was treating functions passed to other bits (ie. the react callbacks) as public facing.

**Comment by dbkr** [comment]
ptal

File: src/components/views/elements/DirectorySearchBox.js (hunk 1 above)
Comment (richvdh) [inline src/components/views/elements/DirectorySearchBox.js]: yeah, but that's bogus :-p.

File: src/components/views/elements/DirectorySearchBox.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        if (this.props.onJoinClick) {
+            this.props.onJoinClick(this.state.value);
+        }
+    }
+
+    _contentLooksLikeAlias() {
+        if (!this.input) return false;
+
+        // liberal test for things that look like room aliases
+        return /^#.+:.+/.test(this.state.value);
```
Comment (richvdh) [inline src/components/views/elements/DirectorySearchBox.js]: trailing `.+` is redundant now

**Pull request merged**